Merge paths can cause extremely large command sequences for a single path

It does not break the image, but in the build processes somewhere (AAPT?) the path data string gets replaced with `STRING_TOO_LARGE`. In this case the path data was 159849 chars long.

```
FATAL EXCEPTION: main
Process: com.chickfila.cfaflagship.debug, PID: 17895
android.content.res.Resources$NotFoundException: Drawable com.chickfila.cfaflagship.debug:drawable/ic_rewards_background_member with resource ID #0x7f080286
Caused by: android.content.res.Resources$NotFoundException: File res/drawable/ic_rewards_background_member.xml from drawable resource ID #0x7f080286
	at android.content.res.ResourcesImpl.loadDrawableForCookie(ResourcesImpl.java:913)
	at android.content.res.ResourcesImpl.loadDrawable(ResourcesImpl.java:681)
	at android.content.res.Resources.loadDrawable(Resources.java:1004)
	at android.content.res.Resources.getDrawableForDensity(Resources.java:994)
	at android.content.res.Resources.getDrawable(Resources.java:933)
	at android.content.Context.getDrawable(Context.java:948)
	at androidx.core.content.ContextCompat$Api21Impl.getDrawable(ContextCompat.java:1049)
	at androidx.core.content.ContextCompat.getDrawable(ContextCompat.java:485)
	at androidx.appcompat.widget.ResourceManagerInternal.getDrawable(ResourceManagerInternal.java:147)
	at androidx.appcompat.widget.ResourceManagerInternal.getDrawable(ResourceManagerInternal.java:135)
	at androidx.appcompat.content.res.AppCompatResources.getDrawable(AppCompatResources.java:64)
	at androidx.appcompat.widget.AppCompatImageHelper.setImageResource(AppCompatImageHelper.java:90)
	at androidx.appcompat.widget.AppCompatImageView.setImageResource(AppCompatImageView.java:102)
	at com.chickfila.cfaflagship.debug.activities.IconTesterActivity$IconsAdapter.bindView(IconTesterActivity.kt:85)
	at com.chickfila.cfaflagship.debug.activities.IconTesterActivity$IconsAdapter.getView(IconTesterActivity.kt:77)
	at android.widget.AbsListView.obtainView(AbsListView.java:2466)
	at android.widget.ListView.makeAndAddView(ListView.java:2065)
	at android.widget.ListView.fillDown(ListView.java:791)
	at android.widget.ListView.fillGap(ListView.java:754)
	at android.widget.AbsListView.trackMotionScroll(AbsListView.java:5557)
	at android.widget.ListView.trackMotionScroll(ListView.java:1982)
	at android.widget.AbsListView$FlingRunnable.run(AbsListView.java:5099)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1339)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1348)
	at android.view.Choreographer.doCallbacks(Choreographer.java:952)
	at android.view.Choreographer.doFrame(Choreographer.java:878)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1322)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8177)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
Caused by: java.lang.IllegalArgumentException: R is not a valid verb. Failure occurred at position 2 of path: STRING_TOO_LARGE
	at android.util.PathParser.nCreatePathDataFromString(Native Method)
	at android.util.PathParser.-$$Nest$smnCreatePathDataFromString(Unknown Source:0)
	at android.util.PathParser$PathData.<init>(PathParser.java:79)
	at android.graphics.drawable.VectorDrawable$VFullPath.updateStateFromTypedArray(VectorDrawable.java:2028)
	at android.graphics.drawable.VectorDrawable$VFullPath.inflate(VectorDrawable.java:1979)
	at android.graphics.drawable.VectorDrawable.inflateChildElements(VectorDrawable.java:840)
	at android.graphics.drawable.VectorDrawable.inflate(VectorDrawable.java:738)
	at android.graphics.drawable.DrawableInflater.inflateFromXmlForDensity(DrawableInflater.java:141)
	at android.graphics.drawable.Drawable.createFromXmlInnerForDensity(Drawable.java:1394)
	at android.graphics.drawable.Drawable.createFromXmlForDensity(Drawable.java:1355)
```

--- a/changelog.md
+++ b/changelog.md
@@ -23,6 +23,7 @@
 - Conversions without a specified output file will write a file the file extension corresponding to the format.
 - Decimal separators are locale-invariant.
 - Crash when using the cli to convert an svg containing a clip path to vector drawable.
+- (Vector Drawable) Path merging avoids merging a single path data string beyond the framework string length limit (#82)
 
 ### Security
 

--- a/changelog.md
+++ b/changelog.md
@@ -4,13 +4,14 @@
 
 ### Added
 - `com.jzbrooks.vgo.core.util.math.Surveyor`, which computes the bounding box of an arbitrary list of commands 
-- Bezier curve interpolation for all variants and elliptical arc bounding box functions
+- Bézier curve interpolation for all variants and elliptical arc bounding box functions
 
 ### Changed
 
 - `vgo-plugin` (`com.jzbrooks.vgo.plugin`) no longer requires a particular version of Android Gradle Plugin.
   Note: `:vgo` is an abstract implementation of the tool which does not assume either a cli or plugin context. CLI related logic has been relocated into `:vgo-cli`.
 - **Breaking:** `CubicCurve<*>.interpolate` has been split into `CubicBezierCurve.interpolate` and `SmoothCubicBezierCurve.interpolate`
+- `com.jzbrooks.vgo.core.optimization.MergePaths` constructor accepts constraints. See `com.jzbrooks.vgo.core.optimization.MergePaths.Constraints`.
 - Paths with an even odd fill rule can be merged
 
 ### Deprecated
@@ -22,7 +23,7 @@
 - Overlapping paths are no longer merged, which avoids some image warping issues (#88, #101)
 - Conversions without a specified output file will write a file the file extension corresponding to the format.
 - Decimal separators are locale-invariant.
-- Crash when using the cli to convert an svg containing a clip path to vector drawable.
+- Crash when using the CLI to convert an SVG containing a clip path to vector drawable.
 - (Vector Drawable) Path merging avoids merging a single path data string beyond the framework string length limit (#82)
 
 ### Security

--- a/vgo-core/api/vgo-core.api
+++ b/vgo-core/api/vgo-core.api
@@ -545,11 +545,36 @@ public final class com/jzbrooks/vgo/core/optimization/ConvertCurvesToArcs : com/
 
 public final class com/jzbrooks/vgo/core/optimization/MergePaths : com/jzbrooks/vgo/core/optimization/BottomUpOptimization {
 	public fun <init> ()V
+	public fun <init> (Lcom/jzbrooks/vgo/core/optimization/MergePaths$Constraints;)V
+	public synthetic fun <init> (Lcom/jzbrooks/vgo/core/optimization/MergePaths$Constraints;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
 	public fun visit (Lcom/jzbrooks/vgo/core/graphic/ClipPath;)V
 	public fun visit (Lcom/jzbrooks/vgo/core/graphic/Extra;)V
 	public fun visit (Lcom/jzbrooks/vgo/core/graphic/Graphic;)V
 	public fun visit (Lcom/jzbrooks/vgo/core/graphic/Group;)V
 	public fun visit (Lcom/jzbrooks/vgo/core/graphic/Path;)V
+}
+
+public abstract interface class com/jzbrooks/vgo/core/optimization/MergePaths$Constraints {
+}
+
+public final class com/jzbrooks/vgo/core/optimization/MergePaths$Constraints$None : com/jzbrooks/vgo/core/optimization/MergePaths$Constraints {
+	public static final field INSTANCE Lcom/jzbrooks/vgo/core/optimization/MergePaths$Constraints$None;
+	public fun equals (Ljava/lang/Object;)Z
+	public fun hashCode ()I
+	public fun toString ()Ljava/lang/String;
+}
+
+public final class com/jzbrooks/vgo/core/optimization/MergePaths$Constraints$PathLength : com/jzbrooks/vgo/core/optimization/MergePaths$Constraints {
+	public fun <init> (Lcom/jzbrooks/vgo/core/graphic/command/CommandPrinter;I)V
+	public final fun component1 ()Lcom/jzbrooks/vgo/core/graphic/command/CommandPrinter;
+	public final fun component2 ()I
+	public final fun copy (Lcom/jzbrooks/vgo/core/graphic/command/CommandPrinter;I)Lcom/jzbrooks/vgo/core/optimization/MergePaths$Constraints$PathLength;
+	public static synthetic fun copy$default (Lcom/jzbrooks/vgo/core/optimization/MergePaths$Constraints$PathLength;Lcom/jzbrooks/vgo/core/graphic/command/CommandPrinter;IILjava/lang/Object;)Lcom/jzbrooks/vgo/core/optimization/MergePaths$Constraints$PathLength;
+	public fun equals (Ljava/lang/Object;)Z
+	public final fun getCommandPrinter ()Lcom/jzbrooks/vgo/core/graphic/command/CommandPrinter;
+	public final fun getMaxLength ()I
+	public fun hashCode ()I
+	public fun toString ()Ljava/lang/String;
 }
 
 public abstract interface class com/jzbrooks/vgo/core/optimization/Optimization {

--- a/vgo-core/src/main/kotlin/com/jzbrooks/vgo/core/optimization/MergePaths.kt
+++ b/vgo-core/src/main/kotlin/com/jzbrooks/vgo/core/optimization/MergePaths.kt
@@ -7,13 +7,16 @@ import com.jzbrooks.vgo.core.graphic.Extra
 import com.jzbrooks.vgo.core.graphic.Graphic
 import com.jzbrooks.vgo.core.graphic.Group
 import com.jzbrooks.vgo.core.graphic.Path
+import com.jzbrooks.vgo.core.graphic.command.CommandPrinter
 import com.jzbrooks.vgo.core.util.math.Surveyor
 import com.jzbrooks.vgo.core.util.math.intersects
 
 /**
  * Merges multiple paths into a single path where possible
  */
-class MergePaths : BottomUpOptimization {
+class MergePaths(
+    private val constraints: Constraints,
+) : BottomUpOptimization {
     private val surveyor = Surveyor()
 
     override fun visit(graphic: Graphic) = merge(graphic)
@@ -52,7 +55,13 @@ class MergePaths : BottomUpOptimization {
         element.elements = elements
     }
 
-    private fun merge(paths: List<Path>): List<Path> {
+    private fun merge(paths: List<Path>): List<Path> =
+        when (constraints) {
+            is Constraints.PathLength -> mergeConstrained(paths, constraints)
+            Constraints.None -> mergeUnconstrained(paths)
+        }
+
+    private fun mergeUnconstrained(paths: List<Path>): List<Path> {
         if (paths.isEmpty()) return emptyList()
 
         val mergedPaths = ArrayList<Path>(paths.size)
@@ -61,10 +70,7 @@ class MergePaths : BottomUpOptimization {
         for (current in paths.drop(1)) {
             val previous = mergedPaths.last()
 
-            // Intersecting paths can cause problems with path fill rules and with transparency.
-            if (!haveSameAttributes(current, previous) ||
-                surveyor.findBoundingBox(previous.commands) intersects surveyor.findBoundingBox(current.commands)
-            ) {
+            if (unableToMerge(previous, current)) {
                 mergedPaths.add(current)
             } else {
                 previous.commands += current.commands
@@ -73,6 +79,50 @@ class MergePaths : BottomUpOptimization {
 
         return mergedPaths
     }
+
+    private fun mergeConstrained(
+        paths: List<Path>,
+        constraints: Constraints.PathLength,
+    ): List<Path> {
+        if (paths.isEmpty()) return emptyList()
+
+        val mergedPaths = ArrayList<Path>(paths.size)
+        mergedPaths.add(paths.first())
+
+        var pathLength =
+            paths
+                .first()
+                .commands
+                .joinToString("", transform = constraints.commandPrinter::print)
+                .length
+
+        for (current in paths.drop(1)) {
+            val previous = mergedPaths.last()
+
+            val currentLength = current.commands.joinToString("", transform = constraints.commandPrinter::print).length
+            val accumulatedLength = pathLength + currentLength
+
+            if (accumulatedLength >= constraints.maxLength || unableToMerge(previous, current)) {
+                mergedPaths.add(current)
+                pathLength = currentLength
+            } else {
+                previous.commands += current.commands
+                pathLength = accumulatedLength
+            }
+        }
+
+        return mergedPaths
+    }
+
+    // Paths must have the same visual parameters to be merged
+    // Intersecting paths can cause problems with path fill rules and with transparency
+    // If constraints exist on a path, they must be updated
+    private fun unableToMerge(
+        previous: Path,
+        current: Path,
+    ): Boolean =
+        !haveSameAttributes(current, previous) ||
+            surveyor.findBoundingBox(previous.commands) intersects surveyor.findBoundingBox(current.commands)
 
     private fun haveSameAttributes(
         first: Path,
@@ -87,4 +137,13 @@ class MergePaths : BottomUpOptimization {
             first.strokeLineCap == second.strokeLineCap &&
             first.strokeLineJoin == second.strokeLineJoin &&
             first.strokeMiterLimit == second.strokeMiterLimit
+
+    sealed interface Constraints {
+        data class PathLength(
+            val commandPrinter: CommandPrinter,
+            val maxLength: Int,
+        ) : Constraints
+
+        data object None : Constraints
+    }
 }

--- a/vgo-core/src/main/kotlin/com/jzbrooks/vgo/core/optimization/MergePaths.kt
+++ b/vgo-core/src/main/kotlin/com/jzbrooks/vgo/core/optimization/MergePaths.kt
@@ -139,8 +139,10 @@ class MergePaths(
             first.strokeMiterLimit == second.strokeMiterLimit
 
     sealed interface Constraints {
+        /** Constraints the optimization by preventing merging paths beyond a given maximum length */
         data class PathLength(
             val commandPrinter: CommandPrinter,
+            /** The maximum length of a single path */
             val maxLength: Int,
         ) : Constraints
 

--- a/vgo-core/src/main/kotlin/com/jzbrooks/vgo/core/optimization/MergePaths.kt
+++ b/vgo-core/src/main/kotlin/com/jzbrooks/vgo/core/optimization/MergePaths.kt
@@ -15,7 +15,7 @@ import com.jzbrooks.vgo.core.util.math.intersects
  * Merges multiple paths into a single path where possible
  */
 class MergePaths(
-    private val constraints: Constraints,
+    private val constraints: Constraints = Constraints.None,
 ) : BottomUpOptimization {
     private val surveyor = Surveyor()
 

--- a/vgo-core/src/main/kotlin/com/jzbrooks/vgo/core/optimization/MergePaths.kt
+++ b/vgo-core/src/main/kotlin/com/jzbrooks/vgo/core/optimization/MergePaths.kt
@@ -102,7 +102,7 @@ class MergePaths(
             val currentLength = current.commands.joinToString("", transform = constraints.commandPrinter::print).length
             val accumulatedLength = pathLength + currentLength
 
-            if (accumulatedLength >= constraints.maxLength || unableToMerge(previous, current)) {
+            if (accumulatedLength > constraints.maxLength || unableToMerge(previous, current)) {
                 mergedPaths.add(current)
                 pathLength = currentLength
             } else {

--- a/vgo-core/src/test/kotlin/com/jzbrooks/vgo/core/optimization/MergePathsTests.kt
+++ b/vgo-core/src/test/kotlin/com/jzbrooks/vgo/core/optimization/MergePathsTests.kt
@@ -10,6 +10,7 @@ import com.jzbrooks.vgo.core.graphic.Group
 import com.jzbrooks.vgo.core.graphic.command.Command
 import com.jzbrooks.vgo.core.graphic.command.CommandVariant
 import com.jzbrooks.vgo.core.graphic.command.EllipticalArcCurve
+import com.jzbrooks.vgo.core.graphic.command.FakeCommandPrinter
 import com.jzbrooks.vgo.core.graphic.command.LineTo
 import com.jzbrooks.vgo.core.graphic.command.MoveTo
 import com.jzbrooks.vgo.core.graphic.command.QuadraticBezierCurve
@@ -416,5 +417,42 @@ class MergePathsTests {
         traverseBottomUp(graphic) { it.accept(optimization) }
 
         assertThat(graphic::elements).hasSize(2)
+    }
+
+    @Test
+    fun pathLengthConstraints() {
+        val paths =
+            listOf(
+                createPath(
+                    listOf(MoveTo(CommandVariant.ABSOLUTE, listOf(Point(0f, 0f)))),
+                ),
+                createPath(
+                    listOf(MoveTo(CommandVariant.ABSOLUTE, listOf(Point(10f, 10f)))),
+                ),
+                createPath(
+                    listOf(MoveTo(CommandVariant.ABSOLUTE, listOf(Point(40f, 40f)))),
+                ),
+                createPath(
+                    listOf(MoveTo(CommandVariant.ABSOLUTE, listOf(Point(50f, 50f), Point(10f, 10f), Point(20f, 30f), Point(40f, 0f)))),
+                ),
+            )
+
+        val graphic = createGraphic(paths)
+        val optimization = MergePaths(MergePaths.Constraints.PathLength(FakeCommandPrinter(), 16))
+
+        traverseBottomUp(graphic) { it.accept(optimization) }
+
+        assertThat(graphic::elements).containsExactly(
+            createPath(
+                listOf(
+                    MoveTo(CommandVariant.ABSOLUTE, listOf(Point(0f, 0f))),
+                    MoveTo(CommandVariant.ABSOLUTE, listOf(Point(10f, 10f))),
+                    MoveTo(CommandVariant.ABSOLUTE, listOf(Point(40f, 40f))),
+                ),
+            ),
+            createPath(
+                listOf(MoveTo(CommandVariant.ABSOLUTE, listOf(Point(50f, 50f), Point(10f, 10f), Point(20f, 30f), Point(40f, 0f)))),
+            ),
+        )
     }
 }

--- a/vgo-core/src/test/kotlin/com/jzbrooks/vgo/core/optimization/MergePathsTests.kt
+++ b/vgo-core/src/test/kotlin/com/jzbrooks/vgo/core/optimization/MergePathsTests.kt
@@ -40,7 +40,7 @@ class MergePathsTests {
             )
 
         val graphic = createGraphic(paths)
-        val optimization = MergePaths()
+        val optimization = MergePaths(MergePaths.Constraints.None)
 
         traverseBottomUp(graphic) { it.accept(optimization) }
 
@@ -90,7 +90,7 @@ class MergePathsTests {
 
         val group = Group(paths)
         val graphic = createGraphic(listOf(group))
-        val optimization = MergePaths()
+        val optimization = MergePaths(MergePaths.Constraints.None)
 
         traverseBottomUp(graphic) { it.accept(optimization) }
 
@@ -161,7 +161,7 @@ class MergePathsTests {
             )
 
         val graphic = createGraphic(paths)
-        val optimization = MergePaths()
+        val optimization = MergePaths(MergePaths.Constraints.None)
 
         traverseBottomUp(graphic) { it.accept(optimization) }
 
@@ -229,7 +229,7 @@ class MergePathsTests {
             )
 
         val graphic = createGraphic(paths)
-        val optimization = MergePaths()
+        val optimization = MergePaths(MergePaths.Constraints.None)
 
         traverseBottomUp(graphic) { it.accept(optimization) }
 
@@ -291,7 +291,7 @@ class MergePathsTests {
             )
 
         val graphic = createGraphic(paths)
-        val optimization = MergePaths()
+        val optimization = MergePaths(MergePaths.Constraints.None)
 
         traverseBottomUp(graphic) { it.accept(optimization) }
 
@@ -411,7 +411,7 @@ class MergePathsTests {
             )
 
         val graphic = createGraphic(listOf(firstHeart, offsetHeart))
-        val optimization = MergePaths()
+        val optimization = MergePaths(MergePaths.Constraints.None)
 
         traverseBottomUp(graphic) { it.accept(optimization) }
 

--- a/vgo/src/main/kotlin/com/jzbrooks/vgo/svg/SvgOptimizationRegistry.kt
+++ b/vgo/src/main/kotlin/com/jzbrooks/vgo/svg/SvgOptimizationRegistry.kt
@@ -21,7 +21,7 @@ class SvgOptimizationRegistry :
                 BakeTransformations(),
                 CollapseGroups(),
                 RemoveEmptyGroups(),
-                MergePaths(),
+                MergePaths(MergePaths.Constraints.None),
             ),
         topDownOptimizations =
             listOf(

--- a/vgo/src/main/kotlin/com/jzbrooks/vgo/vd/VectorDrawableOptimizationRegistry.kt
+++ b/vgo/src/main/kotlin/com/jzbrooks/vgo/vd/VectorDrawableOptimizationRegistry.kt
@@ -21,7 +21,8 @@ class VectorDrawableOptimizationRegistry :
                 BakeTransformations(),
                 CollapseGroups(),
                 RemoveEmptyGroups(),
-                MergePaths(MergePaths.Constraints.PathLength(VectorDrawableCommandPrinter(3), 800)),
+                // https://cs.android.com/android/platform/superproject/main/+/2e48e15a8097916063eacc023044bc90bb93c73e:frameworks/base/libs/androidfw/StringPool.cpp;l=328
+                MergePaths(MergePaths.Constraints.PathLength(VectorDrawableCommandPrinter(3), (1 shl 15) - 1)),
             ),
         topDownOptimizations =
             listOf(

--- a/vgo/src/main/kotlin/com/jzbrooks/vgo/vd/VectorDrawableOptimizationRegistry.kt
+++ b/vgo/src/main/kotlin/com/jzbrooks/vgo/vd/VectorDrawableOptimizationRegistry.kt
@@ -21,7 +21,7 @@ class VectorDrawableOptimizationRegistry :
                 BakeTransformations(),
                 CollapseGroups(),
                 RemoveEmptyGroups(),
-                MergePaths(),
+                MergePaths(MergePaths.Constraints.PathLength(VectorDrawableCommandPrinter(3), 800)),
             ),
         topDownOptimizations =
             listOf(

